app/code/Magento/Catalog/Model/Product/Link/SaveHandler.php

```diff
@@ -55,17 +64,71 @@ public function execute($entityType, $entity)
     {
         $link = $entity->getData($this->metadataPool->getMetadata($entityType)->getLinkField());
         if ($this->linkResource->hasProductLinks($link)) {
-            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity*/
+            /** @var \Magento\Catalog\Api\Data\ProductInterface $entity */
             foreach ($this->productLinkRepository->getList($entity) as $link) {
                 $this->productLinkRepository->delete($link);
             }
         }
         $productLinks = $entity->getProductLinks();
+
+        // Build links per type
+        $linksByType = [];
+        foreach ($productLinks as $link) {
+            $linksByType[$link->getLinkType()][] = $link;
+        }
+
+        // Do check
+        $hasPositionLinkType = $this->isPositionSet($linksByType);
+
+        // Set array position as a fallback position if necessary
+        foreach ($hasPositionLinkType as $linkType => $hasPosition) {
+            if (!$hasPosition) {
+                array_walk($linksByType[$linkType], function ($productLink, $position) {
+                    $productLink->setPosition(++$position);
+                });
+            }
+        }
+
+        // Flatten multi-dimensional linksByType in ProductLinks
+        $productLinks = array_reduce($linksByType, 'array_merge', []);
+
         if (count($productLinks) > 0) {
             foreach ($entity->getProductLinks() as $link) {
                 $this->productLinkRepository->save($link);
             }
         }
         return $entity;
     }
+
+    /**
+     * Check if the position is set for all product links per link type.
+     * array with boolean per type
+     *
+     * @param $linksByType
+     * @return array
+     */
+    private function isPositionSet($linksByType)
```

Comment:
Thinking some more about it, it's confusing that a method name starting with `is` returns an array. It suggests the result should be boolean.

Reply:
Naming convention will be improved